File: leetcode/easy/520_detect_capital/dduck-520_detect_capital.js
// 520. Detect Capital
// https://leetcode.com/problems/detect-capital/

// Runtime: 52 ms, faster than 97.27% of JavaScript online submissions for Detect Capital.
// Memory Usage: 34.2 MB, less than 100.00% of JavaScript online submissions for Detect Capital.
// Runtime: 64 ms, faster than 54.09% of JavaScript online submissions for Detect Capital.
// Memory Usage: 34.3 MB, less than 75.00% of JavaScript online submissions for Detect Capital.
// Runtime: 60 ms, faster than 74.55% of JavaScript online submissions for Detect Capital.
// Memory Usage: 34.1 MB, less than 100.00% of JavaScript online submissions for Detect Capital.

const detectCapitalUse = (word) => {
  return word.match(/([A-Z][a-z]+$)|(^[A-Z]+$)|(^[a-z]*)/)[0] === word;
};

console.log(detectCapitalUse('USA'));
console.log(detectCapitalUse('leetcode'));
console.log(detectCapitalUse('Google'));
console.log(detectCapitalUse('FlaG'));

File: leetcode/easy/520_detect_capital/dduck-520_detect_capital.py
# 520. Detect Capital
# https://leetcode.(com/problems/detect-capital/ (

# Runtime: 32 ms, faster than 36.50% of Python3 online submissions for Detect Capital.
# Memory Usage: 14 MB, less than 6.67% of Python3 online submissions for Detect Capital.
# Runtime: 36 ms, faster than 6.86% of Python3 online submissions for Detect Capital.
# Memory Usage: 13.6 MB, less than 6.67% of Python3 online submissions for Detect Capital.
# Runtime: 36 ms, faster than 6.86% of Python3 online submissions for Detect Capital.
# Memory Usage: 13.5 MB, less than 6.67% of Python3 online submissions for Detect Capital.

from unittest import TestCase

import re


class Solution:
    def detectCapitalUse(self, word: str) -> bool:
        return re.search('([A-Z][a-z]+$)|(^[A-Z]+$)|(^[a-z]*)', word).group(0) == word


TestCase().assertEqual(Solution().detectCapitalUse('USA'), True)
TestCase().assertEqual(Solution().detectCapitalUse('leetcode'), True)
TestCase().assertEqual(Solution().detectCapitalUse('Google'), True)
TestCase().assertEqual(Solution().detectCapitalUse('FlaG'), False)
